clang/lib/Parse/ParseExpr.cpp

```diff
@@ -2324,6 +2327,45 @@ Parser::ParseExprAfterUnaryExprOrTypeTrait(const Token &OpTok,
   return Operand;
 }
 
+/// Parse a __builtin_unique_stable_name expression.  Accepts a type-id or an
+/// arbitrary expression as a parameter.
+ExprResult Parser::ParseUniqueStableNameExpression() {
+  assert(Tok.is(tok::kw___builtin_unique_stable_name) &&
+         "Not __bulitin_unique_stable_name");
+
+  SourceLocation OpLoc = ConsumeToken();
+  BalancedDelimiterTracker T(*this, tok::l_paren);
+
+  // typeid expressions are always parenthesized.
+  if (T.expectAndConsume(diag::err_expected_lparen_after,
+                         "__builtin_unique_stable_name"))
+    return ExprError();
+
+  if (isTypeIdInParens()) {
+    TypeResult Ty = ParseTypeName();
```

Comment:
hehehe, I think we added a new fun thing:
```
int size = 1;
__builtin_unique_stable_name(int[size++]);
```
whee!

Reply:
Did we?  If its a type, we consume close, then return the type.  I would think the `[` character would cause an error on consumeClose.  Good thing to add to a test however...

Reply:
Oh, I don't think this will cause errors, I think this will surprise people that `size` really is incremented even though `__builtin_unique_stable_name` is usually an unevaluated context. This is the same joy that we get from `sizeof`: https://godbolt.org/z/qnb5shaer

Reply:
Woah!  Thats wacky!  And GCC has the same behavior...  That doesn't seem right...

Reply:
That's VLAs for you. C2x 6.5.3.4p2 "If the type of the operand is a variable length array type, the operand is evaluated; otherwise the operand is not evaluated and the result is an integer constant."

Reply:
gross.  

/EOM